File: src/cli/postprocess/export-raw.ts
import path from 'node:path';
import type { LegacyCheckpointRecord, RawNormalizedRecord } from '../../postprocess/types';
import { fromCheckpointRecord, fromDebugRun, toTaskKeyFromDebug } from '../../postprocess/raw-contract';
import { readJsonFile, readJsonLinesFile, timestampForFilename, writeJsonFile, writeJsonLinesFile } from '../../postprocess/io';
import { PATHS } from '../../config/paths';

type CliArgs = {
  checkpointDir?: string;
  debugFile?: string;
  inputJsonl?: string;
  outputJsonl: string;
  outputSummary: string;
};

function printHelp(): void {
  console.log(`Usage:
  npm run postprocess:raw -- [options]

Options:
  --checkpoint-dir=<path>   Checkpoint directory (uses runs.jsonl or raw.jsonl)
  --debug-file=<path>       Build raw records from debug snapshot file
  --input-jsonl=<path>      Re-read canonical raw JSONL and de-duplicate by task_key
  --output-jsonl=<path>     Output raw JSONL (default: artifacts/postprocess/raw.jsonl)
  --output-summary=<path>   Output summary JSON (default: artifacts/postprocess/raw.summary.json)
  -h, --help                Show this help

Examples:
  npm run postprocess:raw
  npm run postprocess:raw -- --checkpoint-dir=artifacts/checkpoints/full-2026-03-14
  npm run postprocess:raw -- --debug-file=artifacts/latest.debug.json
`);
}

function wantsHelp(argv: string[]): boolean {
  return argv.includes('--help') || argv.includes('-h');
}

function parseArgs(argv: string[]): CliArgs {
  const out: CliArgs = {
    outputJsonl: PATHS.postprocess.rawJsonl,
    outputSummary: PATHS.postprocess.rawSummary,
  };

  for (const arg of argv) {
    if (arg.startsWith('--checkpoint-dir=')) {
      const value = arg.split('=')[1] ?? '';
      if (value.trim()) out.checkpointDir = path.resolve(process.cwd(), value.trim());
      continue;
    }
    if (arg.startsWith('--debug-file=')) {
      const value = arg.split('=')[1] ?? '';
      if (value.trim()) out.debugFile = path.resolve(process.cwd(), value.trim());
      continue;
    }
    if (arg.startsWith('--input-jsonl=')) {
      const value = arg.split('=')[1] ?? '';
      if (value.trim()) out.inputJsonl = path.resolve(process.cwd(), value.trim());
      continue;
    }
    if (arg.startsWith('--output-jsonl=')) {
      const value = arg.split('=')[1] ?? '';
      if (value.trim()) out.outputJsonl = path.resolve(process.cwd(), value.trim());
      continue;
    }
    if (arg.startsWith('--output-summary=')) {
      const value = arg.split('=')[1] ?? '';
      if (value.trim()) out.outputSummary = path.resolve(process.cwd(), value.trim());
    }
  }

  if (!out.checkpointDir && !out.debugFile) {
    out.checkpointDir = PATHS.checkpoint.root;
  }

  return out;
}

async function loadFromCheckpoint(checkpointDir: string): Promise<RawNormalizedRecord[]> {
  const runsPath = path.resolve(checkpointDir, 'runs.jsonl');
  const lines = await readJsonLinesFile<LegacyCheckpointRecord>(runsPath);
  const latestByTask = new Map<string, LegacyCheckpointRecord>();

  for (const record of lines) {
    if (record?.task_key && record.metrics && record.debug) {
      latestByTask.set(record.task_key, record);
    }
  }

  return Array.from(latestByTask.values()).map(fromCheckpointRecord);
}

async function loadFromDebug(debugFile: string): Promise<RawNormalizedRecord[]> {
  const debug = await readJsonFile<{ runs?: LegacyCheckpointRecord['debug'][] }>(debugFile);
  const runs = Array.isArray(debug.runs) ? debug.runs : [];
  const latestByTask = new Map<string, LegacyCheckpointRecord['debug']>();

  for (const run of runs) {
    const taskKey = toTaskKeyFromDebug(run);
    latestByTask.set(taskKey, run);
  }

  return Array.from(latestByTask.values()).map(fromDebugRun);
}

async function loadFromCanonicalRaw(inputJsonl: string): Promise<RawNormalizedRecord[]> {
  const lines = await readJsonLinesFile<RawNormalizedRecord>(inputJsonl);
  const latestByTask = new Map<string, RawNormalizedRecord>();
  for (const line of lines) {
    if (line?.task_key && line.model && line.document && line.runtime && line.payload && line.legacy_metrics) {
      latestByTask.set(line.task_key, line);
    }
  }
  return Array.from(latestByTask.values());
}

async function main(): Promise<void> {
  const argv = process.argv.slice(2);
  if (wantsHelp(argv)) {
    printHelp();
    return;
  }
  const args = parseArgs(argv);

  const checkpointRawJsonl = args.checkpointDir ? path.resolve(args.checkpointDir, 'raw.jsonl') : undefined;

  const rawRecords = args.inputJsonl
    ? await loadFromCanonicalRaw(args.inputJsonl)
    : args.debugFile
      ? await loadFromDebug(args.debugFile)
      : checkpointRawJsonl
        ? await loadFromCanonicalRaw(checkpointRawJsonl).catch(async () => loadFromCheckpoint(args.checkpointDir!))
        : await loadFromCheckpoint(args.checkpointDir!);

  rawRecords.sort((a, b) => a.task_key.localeCompare(b.task_key));

  await writeJsonLinesFile(args.outputJsonl, rawRecords);

  const byModel = new Map<string, number>();
  const byDomain = new Map<string, number>();
  let errored = 0;

  for (const record of rawRecords) {
    byModel.set(record.model.model_label, (byModel.get(record.model.model_label) ?? 0) + 1);
    byDomain.set(record.document.domain, (byDomain.get(record.document.domain) ?? 0) + 1);
    if (record.runtime.error !== null) errored += 1;
  }

  const summary = {
    generated_at: new Date().toISOString(),
    source: args.inputJsonl
      ? { input_jsonl: args.inputJsonl }
      : args.debugFile
        ? { debug_file: args.debugFile }
        : checkpointRawJsonl
          ? { checkpoint_raw_jsonl: checkpointRawJsonl, checkpoint_dir: args.checkpointDir }
          : { checkpoint_dir: args.checkpointDir },
    output_jsonl: args.outputJsonl,
    schema_version: '1.0',
    records: rawRecords.length,
    errored_records: errored,
    models: Array.from(byModel.entries())
      .map(([model_label, count]) => ({ model_label, count }))
      .sort((a, b) => a.model_label.localeCompare(b.model_label)),
    domains: Array.from(byDomain.entries())
      .map(([domain, count]) => ({ domain, count }))
      .sort((a, b) => a.domain.localeCompare(b.domain)),
    build_id: `raw-${timestampForFilename()}`,
  };

  await writeJsonFile(args.outputSummary, summary);

  console.log(`Raw records: ${rawRecords.length}`);
  console.log(`Errors: ${errored}`);
  console.log(`JSONL: ${args.outputJsonl}`);
  console.log(`Summary: ${args.outputSummary}`);
}

main().catch((error: unknown) => {
  console.error(error);
  process.exit(1);
});

File: src/ocr_mini_bench/cli/postprocess/export_raw.py
"""`ocr-bench-export-raw`: rebuild `raw.jsonl` from checkpoint records.

Mirrors `src/cli/postprocess/export-raw.ts`. Sources (in priority order):
  --input-jsonl  : re-read canonical raw JSONL, de-duplicate by task_key
  --debug-file   : build from a `latest.debug.json`-style snapshot
  --checkpoint-dir: default — prefer `raw.jsonl` if present, fall back to `runs.jsonl`
"""

from __future__ import annotations

from datetime import UTC, datetime
from pathlib import Path
from typing import Annotated, Any

import typer

from ...config.paths import PATHS
from ...postprocess.io import (
    read_json_file,
    read_json_lines_file,
    timestamp_for_filename,
    write_json_file,
    write_json_lines_file,
)
from ...postprocess.raw_contract import (
    from_checkpoint_record,
    from_debug_run,
    to_task_key_from_debug,
)
from ...postprocess.types import (
    LegacyCheckpointDebug,
    LegacyCheckpointRecord,
    RawNormalizedRecord,
)

app = typer.Typer(
    help="Rebuild raw.jsonl from checkpoint records.",
    add_completion=False,
    no_args_is_help=False,
)


def _resolve_cwd(value: str) -> Path:
    return (Path.cwd() / value).resolve()


def _load_from_checkpoint(checkpoint_dir: Path) -> list[RawNormalizedRecord]:
    runs_path = checkpoint_dir / "runs.jsonl"
    rows = read_json_lines_file(runs_path)
    latest_by_task: dict[str, LegacyCheckpointRecord] = {}
    for row in rows:
        if isinstance(row, dict) and row.get("task_key") and row.get("metrics") and row.get("debug"):
            record = LegacyCheckpointRecord.model_validate(row)
            latest_by_task[record.task_key] = record
    return [from_checkpoint_record(r) for r in latest_by_task.values()]


def _load_from_debug(debug_file: Path) -> list[RawNormalizedRecord]:
    payload = read_json_file(debug_file)
    runs_field = payload.get("runs") if isinstance(payload, dict) else None
    runs = runs_field if isinstance(runs_field, list) else []
    latest_by_task: dict[str, LegacyCheckpointDebug] = {}
    for raw in runs:
        run = LegacyCheckpointDebug.model_validate(raw)
        key = to_task_key_from_debug(run)
        latest_by_task[key] = run
    return [from_debug_run(r) for r in latest_by_task.values()]


def _load_from_canonical_raw(input_jsonl: Path) -> list[RawNormalizedRecord]:
    rows = read_json_lines_file(input_jsonl)
    latest_by_task: dict[str, RawNormalizedRecord] = {}
    for row in rows:
        if (
            isinstance(row, dict)
            and row.get("task_key")
            and row.get("model")
            and row.get("document")
            and row.get("runtime")
            and row.get("payload")
            and row.get("legacy_metrics")
        ):
            record = RawNormalizedRecord.model_validate(row)
            latest_by_task[record.task_key] = record
    return list(latest_by_task.values())


@app.callback(invoke_without_command=True)
def main(
    checkpoint_dir: Annotated[
        str | None,
        typer.Option("--checkpoint-dir", help="Checkpoint directory."),
    ] = None,
    debug_file: Annotated[
        str | None,
        typer.Option("--debug-file", help="Build from a debug snapshot file."),
    ] = None,
    input_jsonl: Annotated[
        str | None,
        typer.Option("--input-jsonl", help="Re-read canonical raw JSONL."),
    ] = None,
    output_jsonl: Annotated[
        str | None,
        typer.Option("--output-jsonl", help="Output raw JSONL."),
    ] = None,
    output_summary: Annotated[
        str | None,
        typer.Option("--output-summary", help="Output summary JSON."),
    ] = None,
) -> None:
    cp_dir: Path | None = _resolve_cwd(checkpoint_dir) if checkpoint_dir else None
    dbg_file: Path | None = _resolve_cwd(debug_file) if debug_file else None
    in_jsonl: Path | None = _resolve_cwd(input_jsonl) if input_jsonl else None
    out_jsonl: Path = _resolve_cwd(output_jsonl) if output_jsonl else PATHS.postprocess.raw_jsonl
    out_summary: Path = (
        _resolve_cwd(output_summary) if output_summary else PATHS.postprocess.raw_summary
    )

    if cp_dir is None and dbg_file is None and in_jsonl is None:
        cp_dir = PATHS.checkpoint.root

    checkpoint_raw_jsonl: Path | None = (cp_dir / "raw.jsonl") if cp_dir is not None else None

    records: list[RawNormalizedRecord]
    if in_jsonl is not None:
        records = _load_from_canonical_raw(in_jsonl)
    elif dbg_file is not None:
        records = _load_from_debug(dbg_file)
    elif checkpoint_raw_jsonl is not None and checkpoint_raw_jsonl.exists():
        try:
            records = _load_from_canonical_raw(checkpoint_raw_jsonl)
        except Exception:
            assert cp_dir is not None
            records = _load_from_checkpoint(cp_dir)
    else:
        assert cp_dir is not None
        records = _load_from_checkpoint(cp_dir)

    records.sort(key=lambda r: r.task_key)
    write_json_lines_file(out_jsonl, [r.model_dump(mode="json") for r in records])

    by_model: dict[str, int] = {}
    by_domain: dict[str, int] = {}
    errored = 0
    for record in records:
        by_model[record.model.model_label] = by_model.get(record.model.model_label, 0) + 1
        by_domain[record.document.domain] = by_domain.get(record.document.domain, 0) + 1
        if record.runtime.error is not None:
            errored += 1

    now = datetime.now(UTC)
    generated_at = now.strftime("%Y-%m-%dT%H:%M:%S.") + f"{now.microsecond // 1000:03d}Z"

    source: dict[str, Any]
    if in_jsonl is not None:
        source = {"input_jsonl": str(in_jsonl)}
    elif dbg_file is not None:
        source = {"debug_file": str(dbg_file)}
    elif checkpoint_raw_jsonl is not None and checkpoint_raw_jsonl.exists():
        source = {
            "checkpoint_raw_jsonl": str(checkpoint_raw_jsonl),
            "checkpoint_dir": str(cp_dir) if cp_dir else None,
        }
    else:
        source = {"checkpoint_dir": str(cp_dir) if cp_dir else None}

    summary = {
        "generated_at": generated_at,
        "source": source,
        "output_jsonl": str(out_jsonl),
        "schema_version": "1.0",
        "records": len(records),
        "errored_records": errored,
        "models": [
            {"model_label": label, "count": count}
            for label, count in sorted(by_model.items(), key=lambda kv: kv[0])
        ],
        "domains": [
            {"domain": d, "count": c}
            for d, c in sorted(by_domain.items(), key=lambda kv: kv[0])
        ],
        "build_id": f"raw-{timestamp_for_filename(now)}",
    }
    write_json_file(out_summary, summary)

    typer.echo(f"Raw records: {len(records)}")
    typer.echo(f"Errors: {errored}")
    typer.echo(f"JSONL: {out_jsonl}")
    typer.echo(f"Summary: {out_summary}")
